File: docs/english.py
"""
English Dictionary Trie - Python Wrapper

Usage:
    from english import EnglishDictionary

    dict = EnglishDictionary.load('english.wasm')
    suggestions = dict.autocomplete('hel', max_results=10)
    is_word = dict.contains('hello')

Requirements:
    pip install wasmtime
"""

from pathlib import Path
from typing import List, Optional
from wasmtime import Store, Module, Instance, Memory


class EnglishDictionary:
    """A compressed English dictionary using a WebAssembly trie."""

    def __init__(self, instance: Instance, store: Store, memory: Memory):
        self._instance = instance
        self._store = store
        self._memory = memory
        self._exports = instance.exports(store)

    @classmethod
    def load(cls, wasm_path: str) -> "EnglishDictionary":
        """
        Load the dictionary from a WASM file.

        Args:
            wasm_path: Path to english.wasm

        Returns:
            EnglishDictionary instance
        """
        store = Store()
        module = Module.from_file(store.engine, wasm_path)
        instance = Instance(store, module, [])

        exports = instance.exports(store)
        memory = exports["memory"]

        # Initialize the trie
        init_fn = exports["init"]
        init_fn(store)

        return cls(instance, store, memory)

    @property
    def node_count(self) -> int:
        """Get the number of nodes in the trie."""
        get_node_count = self._exports["getNodeCount"]
        return get_node_count(self._store)

    def contains(self, word: str) -> bool:
        """
        Check if a word exists in the dictionary.

        Args:
            word: Word to check

        Returns:
            True if the word exists, False otherwise
        """
        query = word.lower().encode('utf-8')
        ptr = self._exports["getResultPtr"](self._store)

        # Write query to memory
        mem_data = self._memory.data_ptr(self._store)
        for i, b in enumerate(query):
            mem_data[ptr + i] = b

        result = self._exports["contains"](self._store, ptr, len(query))
        return bool(result)

    def autocomplete(self, prefix: str, max_results: int = 10) -> List[str]:
        """
        Get autocomplete suggestions for a prefix.

        Args:
            prefix: Prefix to search for
            max_results: Maximum number of results (default: 10)

        Returns:
            List of matching words
        """
        query = prefix.lower().encode('utf-8')
        ptr = self._exports["getResultPtr"](self._store)

        # Write query to memory
        mem_data = self._memory.data_ptr(self._store)
        for i, b in enumerate(query):
            mem_data[ptr + i] = b

        result_len = self._exports["autocomplete"](
            self._store, ptr, len(query), max_results
        )

        if result_len == 0:
            return []

        # Read results from memory
        result_bytes = bytes(mem_data[ptr:ptr + result_len])
        result_text = result_bytes.decode('utf-8')
        return [w for w in result_text.split('\n') if w]


if __name__ == "__main__":
    import sys

    wasm_path = sys.argv[1] if len(sys.argv) > 1 else "english.wasm"

    print(f"Loading dictionary from {wasm_path}...")
    dictionary = EnglishDictionary.load(wasm_path)
    print(f"Loaded {dictionary.node_count:,} nodes")

    # Test autocomplete
    test_prefixes = ["hel", "th", "cat"]
    for prefix in test_prefixes:
        results = dictionary.autocomplete(prefix, 5)
        print(f"\n'{prefix}' -> {results}")

    # Test contains
    test_words = ["hello", "world", "xyznotaword"]
    for word in test_words:
        exists = dictionary.contains(word)
        print(f"\ncontains('{word}') = {exists}")

File: docs/english.rs
//! English Dictionary Trie - Rust Wrapper
//!
//! Usage:
//! ```rust
//! use english::EnglishDictionary;
//!
//! let dict = EnglishDictionary::load("english.wasm")?;
//! let suggestions = dict.autocomplete("hel", 10);
//! let is_word = dict.contains("hello");
//! ```
//!
//! Add to Cargo.toml:
//! ```toml
//! [dependencies]
//! wasmtime = "19"
//! ```

use std::path::Path;
use wasmtime::*;

pub struct EnglishDictionary {
    store: Store<()>,
    memory: Memory,
    init: TypedFunc<(), ()>,
    get_node_count: TypedFunc<(), i32>,
    get_result_ptr: TypedFunc<(), i32>,
    autocomplete: TypedFunc<(i32, i32, i32), i32>,
    contains: TypedFunc<(i32, i32), i32>,
}

impl EnglishDictionary {
    /// Load the dictionary from a WASM file.
    pub fn load<P: AsRef<Path>>(wasm_path: P) -> Result<Self> {
        let engine = Engine::default();
        let mut store = Store::new(&engine, ());

        let module = Module::from_file(&engine, wasm_path)?;
        let instance = Instance::new(&mut store, &module, &[])?;

        let memory = instance.get_memory(&mut store, "memory")
            .ok_or_else(|| anyhow::anyhow!("memory export not found"))?;

        let init = instance.get_typed_func::<(), ()>(&mut store, "init")?;
        let get_node_count = instance.get_typed_func::<(), i32>(&mut store, "getNodeCount")?;
        let get_result_ptr = instance.get_typed_func::<(), i32>(&mut store, "getResultPtr")?;
        let autocomplete = instance.get_typed_func::<(i32, i32, i32), i32>(&mut store, "autocomplete")?;
        let contains = instance.get_typed_func::<(i32, i32), i32>(&mut store, "contains")?;

        // Initialize the trie
        init.call(&mut store, ())?;

        Ok(Self {
            store,
            memory,
            init,
            get_node_count,
            get_result_ptr,
            autocomplete,
            contains,
        })
    }

    /// Get the number of nodes in the trie.
    pub fn node_count(&mut self) -> Result<i32> {
        self.get_node_count.call(&mut self.store, ())
    }

    /// Check if a word exists in the dictionary.
    pub fn contains(&mut self, word: &str) -> Result<bool> {
        let query = word.to_lowercase();
        let query_bytes = query.as_bytes();
        let ptr = self.get_result_ptr.call(&mut self.store, ())?;

        // Write query to memory
        self.memory.write(&mut self.store, ptr as usize, query_bytes)?;

        let result = self.contains.call(&mut self.store, (ptr, query_bytes.len() as i32))?;
        Ok(result != 0)
    }

    /// Get autocomplete suggestions for a prefix.
    pub fn autocomplete(&mut self, prefix: &str, max_results: i32) -> Result<Vec<String>> {
        let query = prefix.to_lowercase();
        let query_bytes = query.as_bytes();
        let ptr = self.get_result_ptr.call(&mut self.store, ())?;

        // Write query to memory
        self.memory.write(&mut self.store, ptr as usize, query_bytes)?;

        let result_len = self.autocomplete.call(
            &mut self.store,
            (ptr, query_bytes.len() as i32, max_results),
        )?;

        if result_len == 0 {
            return Ok(Vec::new());
        }

        // Read results from memory
        let mut result_bytes = vec![0u8; result_len as usize];
        self.memory.read(&self.store, ptr as usize, &mut result_bytes)?;

        let result_text = String::from_utf8(result_bytes)?;
        let words: Vec<String> = result_text
            .split('\n')
            .filter(|w| !w.is_empty())
            .map(String::from)
            .collect();

        Ok(words)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_load_and_query() -> Result<()> {
        let mut dict = EnglishDictionary::load("english.wasm")?;

        // Test node count
        let count = dict.node_count()?;
        assert!(count > 0);

        // Test contains
        assert!(dict.contains("hello")?);
        assert!(!dict.contains("xyznotaword")?);

        // Test autocomplete
        let results = dict.autocomplete("hel", 5)?;
        assert!(!results.is_empty());

        Ok(())
    }
}
